homeassistant/components/ted5000/sensor.py

```diff
@@ -49,85 +165,148 @@ def setup_platform(
 
-    def __init__(self, gateway, name, mtu, unit):
-        """Initialize the sensor."""
-        units = {POWER_WATT: "power", ELECTRIC_POTENTIAL_VOLT: "voltage"}
-        self._gateway = gateway
-        self._name = f"{name} mtu{mtu} {units[unit]}"
-        self._mtu = mtu
-        self._unit = unit
-        self.update()
+def get_ted5000(self) -> int | None:
+    """Collect data from the ted5000 gateway."""
+
+    url = self.url
+    self.data = {}
+    self.data_utility = {}
+
+    try:
+        request = requests.get(url, timeout=10)
+    except requests.exceptions.RequestException as err:
+        _LOGGER.error("No connection to endpoint: %s", err)
+    else:
+        doc = xmltodict.parse(request.text)
+        mtus = int(doc["LiveData"]["System"]["NumberMTU"])
 
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return self._name
+        # MTU Data
+        for mtu in range(1, mtus + 1):
+            power = int(doc["LiveData"]["Power"]["MTU%d" % mtu]["PowerNow"])
+            voltage = int(doc["LiveData"]["Voltage"]["MTU%d" % mtu]["VoltageNow"])
+            energy_tdy = int(doc["LiveData"]["Power"]["MTU%d" % mtu]["PowerTDY"])
+            energy_mtd = int(doc["LiveData"]["Power"]["MTU%d" % mtu]["PowerMTD"])
+            power_factor = int(doc["LiveData"]["Power"]["MTU%d" % mtu]["PF"])
 
-    @property
-    def native_unit_of_measurement(self):
-        """Return the unit the value is expressed in."""
-        return self._unit
+            self.data[mtu] = {
+                ENTITY_MTU_POWER: power,
+                ENTITY_MTU_VOLTAGE: voltage / 10,
+                ENTITY_MTU_TODAY: energy_tdy / 1000,
+                ENTITY_MTU_MONTH: energy_mtd / 1000,
+                ENTITY_MTU_PF: power_factor / 10,
+            }
 
-    @property
-    def native_value(self):
-        """Return the state of the resources."""
-        with suppress(KeyError):
-            return self._gateway.data[self._mtu][self._unit]
+        # Utility Data
+        current_rate = int(doc["LiveData"]["Utility"]["CurrentRate"])
+        days_left = int(doc["LiveData"]["Utility"]["DaysLeftInBillingCycle"])
+        plan_type = int(doc["LiveData"]["Utility"]["PlanType"])
+        plan_type_str = {0: "Flat", 1: "Tier", 2: "TOU", 3: "Tier+TOU"}
+        carbon_rate = int(doc["LiveData"]["Utility"]["CarbonRate"])
+        read_date = int(doc["LiveData"]["Utility"]["MeterReadDate"])
 
-    def update(self) -> None:
-        """Get the latest data from REST API."""
-        self._gateway.update()
+        if plan_type in (0, 2):
+            current_tier = 0
+        else:
+            current_tier = int(doc["LiveData"]["Utility"]["CurrentTier"]) + 1
+
+        if plan_type < 2:
+            current_tou = 0
+            current_tou_str = "Not Configured"
+        else:
+            current_tou = int(doc["LiveData"]["Utility"]["CurrentTOU"]) + 1
+            current_tou_str = doc["LiveData"]["Utility"]["CurrentTOUDescription"]
```

Comment:
I realise that this PR is adding to existing functionality, and largely accessing the data in the same way as it was previously, however this does seem to be reaching quite far into specifics of the ted5000 API.

I would strongly consider moving this into a pypi library that can be imported, as this is what is requested in the 'Development Checklist'.

I believe the reason for the request is to reduce the complexity and quantity of HA PRs relating to API changes.

Maybe it can be accepted for this PR, will need a 2nd opinion from one of the other core devs.

Reply:
It's not reaching far at all since it's simply reading more lines in the XML file returned by the TED5000. It's one single request to access the file then extract multiple lines from it.

As previously stated, I'm NOT a python pro and over struggled for these simple additions, Hopefully it could be accepted as is otherwise I'd rather close this PR and keep it to myself, it's becoming absolutely unsustainable.

Reply:
Sorry to hear that - thanks for persevering so far.  As stated, we can get a 2nd opinion.  It is just using more of the same method of accessing the data, so maybe it is ok.

Reply:
Hopefully a second opinion will rule this out as I don't intend to put more time on this as it works flawlessly for me for over a year now. This PR dates back to October 2021...